Fixes #398 - hooking up two expanders

Jetski5822 opened the pull request:
@jtkech what do you think?

sebastienros commented on src/Orchard.DisplayManagement/Razor/ThemeAwareViewLocationExpander.cs:
```diff
@@ -16,50 +16,67 @@ public class ThemeAwareViewLocationExpander : IViewLocationExpander
-                var extensionViewsPath = Path.Combine(Path.DirectorySeparatorChar
-                    + extension.SubPath, "Views");
+                var site = siteService.GetSiteSettingsAsync().GetAwaiter().GetResult();
+                var adminThemeId = (string)site.Properties["CurrentAdminThemeName"];
```
That's why we should let PRs leave a little, and make some, this is wrong. the property is an implementation detail of a special theme selector.

sebastienros replied:
If we are in the Admin, then the admin theme is the value of the previous `currentTheme`, I don't think we need to special-case the admin theme.

jtkech commented:
@sebastienros 

Sorry again. 

About the admin theme it's because when we know we are in the admin theme, there is a specific implementation, as in O1, where all themes are considered as a base theme of the admin.

Here the comments grabbed from O1.

    // Special case: conceptually, the base themes of "TheAdmin" is the list of all
    // enabled themes. This is so that any enabled theme can have controller/action/views
    // in the Admin of the site.

So, if we really need to know if we are in admin, need more time to see how we can do it in another way. Or maybe we don't want to implement this special case as in O1.

jtkech commented:
@sebastienros 

So, yes i understand about the theme selector, and in the related code in O1 an hard coded string "TheAdmin" is used. It is used to allow a theme to override a controller views in the admin (if not already done by the admin theme itself).

So, if you think we don't need this functionality, maybe just remove the related parts of code.

@Jetski5822 yes i think you're right to have 2 expanders.

Just for infos, before only using `ThemeAwareViewLocationExpander`, The `ModuleViewLocationExpander` was defined in `Microsoft.AspNetCore.Mvc.Modules.Hosting` and added in `AddModuleServices()`.

But it's ok for me that they are both added in `AddThemingHost()`. This because at the end views locations related to themes need to be at the top. So here we would just have to add the theme expander after the module one.

But i wanted to point out where `ModuleViewLocationExpander` was added before.

Last thing, when first fixing the master branch (before i broke it myself) i was registering `IFeaturesProvider` not only in `AddExtensionManagerHost()` but also here as it was done before with `FeatureManager`. Not sure it's important, just to point out.

        // called by Orchard.Commons startup
        public static IServiceCollection AddExtensionManager(this IServiceCollection services)
        {
            services.TryAddScoped<IFeaturesProvider, FeaturesProvider>();
            services.TryAddTransient<IFeatureHash, FeatureHash>();

            return services;
        }

sebastienros commented:
If you don't want to add specific dependencies, we can always add a new interface, like an `IViewLocationProvider`. It would be used in the view location expander and provided by different modules (admin, theming ...). You could then also have a single viewlocationexpander, but different IViewLocationProviders depending on the context (setup, non-setup, ...) Could also have eahc ViewLocationProvier result ordered for priorities.

jtkech commented:
Oh yes, i think i understand, then e.g one expander implemented by `Orchard.Admin` which do more specialized things. Oh yes and a priority to have the right expanders order ...

So, i will continue on views expanders when things will be stabilized, our works with @Jetski5822 was very tied. He found bugs in my code and i had some suggestions on his work. One advantage i see here is that at least we are better aware of what each one needs.

About this, just one thing, what do you think about the following

    // called by Orchard.Commons startup
    public static IServiceCollection AddExtensionManager(this IServiceCollection services)
    {
       // also registered as a singleton in AddExtensionManagerHost()
       // do we need to add this line as used before for FeatureManager?
        services.TryAddScoped<IFeaturesProvider, FeaturesProvider>();
        ...
        return services;
    }

sebastienros commented:
You guys can also work on the same branch, and even do PRs for a working branch. jt you have write access to this repo if you want to create a branch, or work on one of nick's branches.

jtkech commented:
Yes, you're right, need to have another talk with @Jetski5822 to share our last findings and on which branch we can work together.

Jetski5822 commented:
@jtkech Hey man - I pushed my tests and the fixed code. I had to override what you did in my commit.

1. There was a bug in the FeatureBuilder event where the context value for base theme was being set, but not returned. So not Base Theme is a implicit dependency. I included a test for this.

I also recast ExtensionInfo to ThemeExtensionInfo too.

2. All Features returned are ordered by dependency and priority and then name. The Least demanding Feature is returned First. I have added a bunch of tests for this, and this behaviour is the same for all methods when requesting features.

*Next*
- add tests where the Theme has a dependency on a Module
- Add a LocationExpanderProvider, with a composite. Create a raw Expander to take these values.

Jetski5822 commented:
@jtkech

1. added a composite, and extracted the individual selectors out.
2. added a priority and order by descending
3. created a DefaultViewLocationExpanderProvider which just returns the viewLocations
4. Introducted an AdminTheme one, would you mind taking a look?

Also, to get base themes now, you just need to do this...

```
var currentThemeAndBaseThemesOrdered = extensionManager
     .GetFeatures(new[] { currentTheme.Id })
     .Where(x => x.Extension.Manifest.IsTheme());
```

As the base theme is an implicit dependency. Remember my unit test from the previous commit prove this, so this should clean a bunch of things up.

*next*

add tests where the Theme has a dependency on a Module

Jetski5822 commented:
- Hooked up ability to have multiple base themes - was easy, and I added a test for it.

*next*

IExtensionOrderingStrategy.... this will allow us to do the ordering you want outside of the Extension Manager

Jetski5822 commented:
Implemented IExtensionOrderingStrategy, hooked up theme ordering, so Extension Features first, then Themes.

Themes are ordered by either dependencies, or basethemes which in implicit.

*next*

A few more tests

@sebastienros Can you review what ive done please.

jtkech commented:
@Jetski5822 

Wow, awesome work, can't right now but i long to see more in depth.

Thanks for having done all the structure to define order strategies and view locations in other places. Here maybe define and register the modules views expander in Modules.Hosting as before, and the admin theme views expander in Orchard.Admin (e.g not to use the theme selector but the admin theme service).

I think that now all things we need are there and as usual i will disturb you but only on a few points i've seen after a quick look. But again awesome work.

I think there is still a mixing between has dependency and is dependent. E.g see #406 which i could fix (yesterday, not with your current updates) on my machine for the same reason.

And i still think that when just ordering a collection, which already have all features and dependencies, with the ordering utility, we don't need to go through all the dependencies graph for each feature. And we don't need to re-use the ordering utility when we start from a collection which is already ordered.

There were some drawbacks when you tested it but it was the same when i tested it with your own version. So, these drawbacks were there because of other issues i think you have fixed.

Can't right now but this night i'll try to check this, at least i will put my suggestions here.

Note: And for the BaseTheme, because we now have ordering strategies, maybe use this implicit dependency only for specific theme ordering (as in O1), and then without having to use event builders.

Again, awesome work.

Best.

sebastienros commented on src/Orchard.DisplayManagement/LocationExpander/CompositeViewLocationExpanderProvider.cs:
```diff
@@ -0,0 +1,53 @@
+    {
+        public double Priority
+        {
+            get { throw new NotSupportedException(); }
```
dangerous, just return a default value.

sebastienros commented on src/Orchard.DisplayManagement/LocationExpander/AdminThemeAwareViewLocationExpanderProvider.cs:
```diff
@@ -0,0 +1,83 @@
+        }
+
+        /// <inheritdoc />
+        public virtual IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context,
```
Single line if it fits

sebastienros commented on src/Orchard.DisplayManagement/LocationExpander/DefaultViewLocationExpanderProvider.cs:
```diff
@@ -0,0 +1,28 @@
+{
+    public class DefaultViewLocationExpanderProvider : IViewLocationExpanderProvider
+    {
+        public double Priority
```
`public double Priority = 0D;`

Jetski5822 replied:
Done

sebastienros commented on src/Orchard.DisplayManagement/LocationExpander/ModuleViewLocationExpanderProvider.cs:
```diff
@@ -1,15 +1,22 @@
     {
+        public double Priority
+        {
+            get
```
same

Jetski5822 replied:
Done

sebastienros commented on src/Orchard.DisplayManagement/LocationExpander/ThemeAwareViewLocationExpanderProvider.cs:
```diff
@@ -0,0 +1,84 @@
+{
+    public class ThemeAwareViewLocationExpanderProvider : IViewLocationExpanderProvider
+    {
+        public double Priority
```
same

Jetski5822 replied:
Done

sebastienros commented on src/Orchard.Environment.Extensions.Abstractions/IExtensionOrderingStrategy.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IExtensionOrderingStrategy
+    {
+        bool HasDependency(IFeatureInfo observer, IFeatureInfo subject);
```
See comment about name and returned value.

Jetski5822 replied:
When I wrote that name I thought it sucked. I'll change

sebastienros commented on src/Orchard.Environment.Extensions/ExtensionManager.cs:
```diff
@@ -307,5 +321,22 @@ private static string GetSourceFeatureNameForType(Type type, string extensionId)
+
+            public TValue GetOrAdd(TKey key, Func<TKey, TValue> valueFactory)
+            {
+                var lazyResult = this.concurrentDictionary.GetOrAdd(key, k => new Lazy<TValue>(() => valueFactory(k), LazyThreadSafetyMode.ExecutionAndPublication));
```
Why the thread safety option? Either you comment the explicit reason because it is understood and it solves an issue, or you use the default.

Also, can we write it `new Lazy<TValue>(valueFactory)` ?

Jetski5822 replied:
Added comment, its to do with the valuefactory not being thread safe https://blogs.endjin.com/2015/10/using-lazy-and-concurrentdictionary-to-ensure-a-thread-safe-run-once-lazy-loaded-collection/

sebastienros commented on test/Orchard.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs:
```diff
@@ -126,7 +125,7 @@ public TestThemeExtensionInfo(string name, IFeatureInfo baseTheme)
                     {"desciption", name},
                     {"type", "theme"},
-                    {"basetheme", baseTheme.Id}
+                    {"basethemes", baseTheme.Id }
```
Is it to represent the list of base themes in the whole hierarchy? or do you really mean a theme can inherit from two independent ones?

Jetski5822 replied:
Because an extension supports multiple dependencies, a theme is just an extension, and base theme is a implicit dependency referencer. So we can support multiple base themes.

sebastienros commented on test/Orchard.Tests/Extensions/TestDependencyModules/DerivedThemeSample2/Theme.txt:
```diff
@@ -0,0 +1,5 @@
+Author: Nicholas Mayne
+Description: Base Theme
+Version: 1.0.0 
+BaseThemes: BaseThemeSample, BaseThemeSample2
```
That worries me. What's the scenario?

Jetski5822 replied:
Well extensions support it, and was easy to support

sebastienros replied:
BaseTheme and dependency are two things. A theme can depend on other features, and can inherit from one theme, the implied behavior is not the same.

Jetski5822 replied:
BaseThemes and Dependencies are the same terms, there is no difference in a Base Theme/Theme/Module etc etc. From an extensions perspective, they are just extensions.

Themes use the BaseThemes property as a way to say, hey, I depend on you. You could put this in the Dependencies property, and semantically that would be fine.

Now... When I say they are all the same, they only holds true within the Extensions code. Within the Display code, there is a ThemeExtensionInfo that has other things. That I think uses BaseThemes for something.

jtkech commented:
@Jetski5822 

I prefer to wait before testing the code and start with some comments step by step.

---
@sebastienros about your comment on `HasDependency` it is used only for ordering when using the `OrderByDependenciesAndPriorities()` utility which has an `hasDependency` func as a parameter.

Indeed we can change the name, and the parameter name of the utility. This is the name used in the related code in O1 (see below). Notice that only the 1st level of explicit dependencies is checked, not all the graph. The ordering utility do the job when doing the same for all items.

But this assumes that the collection contains all dependencies.

Oh, also notice the case insensitive comparison, maybe for cases as `TinyMCE or TinyMce`. Just to point out if we need to do the same.

    internal static bool HasDependency(FeatureDescriptor item, FeatureDescriptor subject) {
        if (DefaultExtensionTypes.IsTheme(item.Extension.ExtensionType)) {
            if (DefaultExtensionTypes.IsModule(subject.Extension.ExtensionType)) {
                // Themes implicitly depend on modules to ensure build and override ordering
                return true;
            }

            if (DefaultExtensionTypes.IsTheme(subject.Extension.ExtensionType)) {
                // Theme depends on another if it is its base theme
                return item.Extension.BaseTheme == subject.Id;
            }
        }

        // Return based on explicit dependencies
        return item.Dependencies != null &&
               item.Dependencies.Any(x => StringComparer.OrdinalIgnoreCase.Equals(x, subject.Id));
    }

---
@Jetski5822 oh, good idea to only check if it's not a theme in place of checking if it is a core module or a module or a theme ... I thought about this but was not sure we can do it.

        public bool HasDependency(IFeatureInfo observer, IFeatureInfo subject)
        {
            if (observer.Extension.Manifest.IsTheme()) {
                return !subject.Extension.Manifest.IsTheme();
            }
            return false;
        }

That's said, after a quick look in O1, it seems that baseTheme is used in a specific way for placement, templates, when enabling / disabling a theme, e.g you can't have more than one base theme ...

So, it's just a suggestion but maybe better to also implement the baseTheme implicit dependency here, just for ordering purpose. And not have to add it in the dependencies when building a theme feature.

---
For the code below, as already discussed, when applied to a collection which also contains features dependencies, no need to go through all the dependency graph (only the 1st level). The ordering utility will do the job when the same is applied to all items. See above the related code in O1.

But this assumes (and right now this is the case) that the collection contains all dependencies. So, and because it is now cached, **finally i agree with you to use the graph** to be more consistent.

Maybe better to use `GetFeatureDependencies()` to see directly in the feature dependencies in place of parsing all features to grab their dependencies and see if it contains the feature.

So, in place of this.

        public bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
        {
            var isDependent = _extensionOrderingStrategy.HasDependency(f1, f2);

            if (isDependent)
            {
                return true;
            }

            return GetDependentFeatures(f2.Id).Contains(f1);
        }

Beyond the name we can replace, i would use

        public bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
        {
            var hasDependency = _extensionOrderingStrategy.HasDependency(f1, f2);

            if (hasDependency)
            {
                return true;
            }

            return GetFeatureDependencies(f1.Id).Contains(f2);
        }

---
Finally, for the following code.

        public IEnumerable<IFeatureInfo> GetFeatures(string[] featureIdsToLoad)
        {
            var allUnorderedFeaturesToLoadIncludingDependencies = featureIdsToLoad
                .SelectMany(featureId => GetFeatureDependencies(featureId))
                .Distinct();

            return Order(allUnorderedFeaturesToLoadIncludingDependencies);
        }

As discussed, i would start from `GetFeatures()` which is cached and already ordered, and which contains all features grabbed from all available extensions. Then you don't need to re-order things.

        public IEnumerable<IFeatureInfo> GetFeatures(string[] featureIdsToLoad)
        {
            var allDependencies = featureIdsToLoad
                .SelectMany(featureId => GetFeatureDependencies(featureId))
                .Distinct();

            var orderedFeatureDescriptors = GetFeatures()
                .Where(f => allDependencies.Any(d => d.Id == f.Id));

            return orderedFeatureDescriptors;
        }

Best.

jtkech commented:
@sebastienros @Jetski5822 

Finally after re-reading the code, just last things before leaving.

Not sure, but i think that the **HasDepency** name is not so bad because it is intended to be only used for ordering purpose but with the OrderBy**Dependencies**AndPriorities() utility which has 2 func parameters, **hasDependency** and getPriority. So, still related to dependencies (implicit and explicit) but just for ordering.

About mixing order and (explicit) dependencies, it's one reason (see previous comment) that now i think maybe better not to add BaseThemes in explicit dependencies and use them for some specific local implementations and as before for ordering. Here we could use the new theme order strategy.

---
Finally, not sure but i don't really agree that it is wrong to use the **admin theme selector** in the view expander. In fact it is not directly used but compared to the current theme. This to implement a specific **GetBaseThemes** behavior (as in O1) if currentTheme == adminTheme.

In this case, any enabled theme is considered as a base theme of the admin theme. This to allow any enabled theme to override a controller view when in admin.

The code is in the previous version but see below the related code in O1 and its comments where the current theme is compared to the hard coded string "TheAdmin".

        private IEnumerable<ExtensionDescriptor> GetBaseThemes(ExtensionDescriptor themeExtension) {
            if (themeExtension.Id.Equals("TheAdmin", StringComparison.OrdinalIgnoreCase)) {
                // Special case: conceptually, the base themes of "TheAdmin" is the list of all
                // enabled themes. This is so that any enabled theme can have controller/action/views
                // in the Admin of the site.
                return _extensionManager
                    .EnabledFeatures(_shellDescriptor)
                    .Reverse()  // reverse from (C <= B <= A) to (A => B => C)
                    .Select(fd => fd.Extension)
                    .Where(fd => DefaultExtensionTypes.IsTheme(fd.ExtensionType));
            }
            else {
            ...


Note: If we add the admin theme expander in Orchard.Admin we could use the AdminThemeService.

Best

Jetski5822 commented:
@sebastienros @jtkech pushed a bunch of fixes.

JT: I dont understand the Admin thing you were talking about, can you elaborate? the whole GetBaseThemes thing

sebastienros commented:
@jtkech 

> his to allow any enabled theme to override a controller view when in admin.

Then this one is the current theme at the time the view is rendered, and it will be first in the ordered features, so I confirm my opinion we should not know about `admin` in the expander.

> HasDependency name

Features are ordered. There can be many ways to order them:
1. Dependencies, as in "This module can't work without this other module"
2. Dependencies as in "this module needs to run in a deterministic order compared to this other"
3 Type of module: theme vs module
4 Priority metadata
5. Whatever other reason someone could find as we are making it extensible

So why call is `HasDependency` ? Let's call it Order, Compare, ShouldIRunItFirst?
And if the name `OrderByDependenciesAndPriorities` is wrong, let's not drive the implementation by the name of this method, but change the name to `OrderFeatures` for instance ;)

jtkech commented:
@Jetski5822 

Thanks to take into account my comments, as usual i can't check it right now, will do this night.

Thanks for having done the job in shape table mananger which don't need anymore extra code to order by dependencies.

Good to have moved admin theme views expander to Orchard.Admin. Maybe good to also move the module views expander in `Microsoft.AspNetCore.Mvc.Modules.Hosting/Mvc/Razor` where it was originally before i did my 1st theme views expander in `Orchard.DisplayManagement`.

---
I think that themes can have regular explicit depencencies on each others, but BaseTheme is used in another way for placements, templates ... So, currently a theme can have only one baseTheme (and this baseTheme can have its own baseTheme).

---
I will elaborate this night about the admin views expander, in a few words when the admin theme is the current theme, all enabled themes are considered as the base theme of the admin theme. Otherwise do nothing here.

But you don't need to implement this internally, this is a special case that need to be implemented locally in the admin views expander. See my 1st implementation of views expander.

https://github.com/OrchardCMS/Orchard2/blob/892c21a938c1ae14a7c4005354a482d3dc2a8178/src/Orchard.DisplayManagement/Razor/ThemeAwareViewLocationExpander.cs

Will give more precise details this night.

---
Oh, right now just one thing, e.g in the following code. To know if a feature f1 has a dependency (hasDependency) on f2, it's more natural for me to use `GetFeatureDependencies()(f1.Id).Contains(f2)`. This in place of parsing features which have a dependency on f2, and then see if it contains f1 (f1 isDependent), this by using `GetDependentFeatures(f2.Id).Contains(f1)`.

        public bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
        {
            var isDependent = _extensionOrderingStrategy.HasDependency(f1, f2);

            if (isDependent)
            {
                return true;
            }

            return GetDependentFeatures(f2.Id).Contains(f1);
        }


Best.

jtkech commented:
@sebastienros just seen your comment while answering to @Jetski5822 (see above) 

So, yes for the name, it was just to point out the other related names we would need to also change.

> Then this one is the current theme at the time the view is rendered, and it will be first in the ordered features, so I confirm my opinion we should not know about admin in the expander.

Yes, it will be the first in the list, but after we need to know which "base themes" to add, in a normal case we add the theme.BaseTheme (and the baseTheme of the baseTheme ...). But in this special case we add all enabled themes. So any enabled theme can override an admin controller view (if not already done by the admin theme itself).

As usual, need to leave at this time, will elaborate this night.

Best.

jtkech commented:
@sebastienros about the admin theme.

So, for theme views locations expander we use ordered features (without modules) and according to the current theme. So we use the following list.

    currentTheme
    regular baseTheme1 of currentTheme
    regular baseTheme2 of baseTheme1
    ...

When the currentTheme == adminTheme, the specific beharvior (as in O1) is to consider all enabled themes as a baseTheme of the adminTheme. Easy to do but need to be done only in this context.

    currentTheme (adminTheme) // first as above
    enabled theme1 // considered as a baseTheme in this context
    enabled theme2 // idem
    ...

Note: this behavior doesn't need to be implemented internally, only locally in this context.

Best.

jtkech commented:
@Jetski5822 

- Just fetched your work on my machine. As said, now all we need seems to be there, again i learned a lot on how to share implementations between projects, thanks.

- About the baseTheme i'm not against at all to use it also as a regular dependency, just a little afraid of the impact. Need more time to see specific usage in O1 not yet implemented in O2. meanwhile i let you decide with @sebastienros.

- About the admin theme, see my above comments, let me know if you have questions. **But i can do it and the remaining job related to views expanders**, this when you will be ready with your work.

- Then i will be able to check features ordering on my side, and you will be able to review my code, and with @sebastienros for the special case of the admin theme.

So, right now on my side, only one thing disturb me a little. As said above, personnaly i would have used.

    public bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
    {
        var hasDependency = _extensionOrderingStrategy.HasDependency(f1, f2);

        if (hasDependency)
        {
            return true;
        }

        return GetFeatureDependencies(f1.Id).Contains(f2);
    }

This beyond the fact that maybe we need to change all `hasDependency` terms as suggested by @sebastienros. This when used for ordering purpose, i will have a thought on this.

Best.

Jetski5822 commented:
@jtkech Fixed the HasDependency GetDependent/etc thing

jtkech commented:
@Jetski5822 ok, thanks for this one ;)

And it is 2 time faster on 1st execution, but not so important because things are cached.

jtkech commented:
@Jetski5822 

Just tried some tests, will do more tomorrow.

So, all ordering tests i've done work, this by modifying manifest dependencies / priority. First it was not working with the `BaseTheme` because here i needed to use `BaseThemes`. Right? Then it works.

Then shape table manager unit tests was failing, but it is my fault, in the `TestExtensionManager` the `GetFeatureDependencies()` was not yet implemented.

So, in `DefaultShapeTableManagerTests.cs` you just need to replace this

            public IEnumerable<IFeatureInfo> GetFeatureDependencies(string featureId)
            {
                throw new NotImplementedException();
            }
            ...
            ...
            public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
            {
                return _features.Where(x => x.Dependencies.Contains(featureId));
            }

With this

            public IEnumerable<IFeatureInfo> GetFeatureDependencies(string featureId)
            {
                return _features.Where(x => x.Dependencies.Contains(featureId));
            }
            ...
            ...
            public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
            {
                throw new NotImplementedException();
            }

Then all tests pass.

Last thing, shape tables creations seems to be longer. Oh, maybe because of a bunch of new modules / features. Will see tomorrow night.

Best.

Jetski5822 commented:
Hey I have a change to push too :)

Sent from my iPhone

> On 13 Dec 2016, at 11:59, jtkech <notifications@github.com> wrote:
> 
> @Jetski5822
> 
> Just tried some tests, will do more tomorrow.
> 
> So, all ordering tests i've done work, this by modifying manifest dependencies / priority. First it was not working with the BaseTheme because here i needed to use BaseThemes. Right? Then it works.
> 
> Then shape table manager unit tests was failing, but it is my fault, in the TestExtensionManager the GetFeatureDependencies() was not yet implemented.
> 
> So, in DefaultShapeTableManagerTests.cs you just need to replace this
> 
>         public IEnumerable<IFeatureInfo> GetFeatureDependencies(string featureId)
>         {
>             throw new NotImplementedException();
>         }
>         ...
>         ...
>         public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
>         {
>             return _features.Where(x => x.Dependencies.Contains(featureId));
>         }
> With this
> 
>         public IEnumerable<IFeatureInfo> GetFeatureDependencies(string featureId)
>         {
>             return _features.Where(x => x.Dependencies.Contains(featureId));
>         }
>         ...
>         ...
>         public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
>         {
>             throw new NotImplementedException();
>         }
> Then all tests pass.
> 
> Last thing, shape tables creations seems to be longer. Oh, maybe because of a bunch of new modules / features. Will see tomorrow night.
> 
> Best.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jtkech commented:
Oh, do you also mean that you want i push the above fix of unit tests?
Or should i let you do it?

Jetski5822 commented:
Fix for compare, which looks good.

@sebastienros Do you have any last comments on this PR?

jtkech commented:
Oh, you fixed the compare, but it was working in my tests. Need to see it tomorrow.

Just to point out there is this file which is necessary to build because the same in Orchard.Environment.Extensions is declared as internal. Is it temporary?
    /Orchard2/src/Orchard.DisplayManagement/InternalSources/Orchard.Environment.Extensions/CompositeExtensionOrderingStrategy.cs

Oh, about the shape tables creations which take a longer time. It's because of the new order strategy applied to each item when we order shape descriptors (which are tied to features).

But i think i have a solution because we are doing twice the same thing and it is cached on your side. So when in shape table manager we get the ordered features, we can use this list already ordered according to dependencies and priority, and e.g just use the index of each feature as a pseudo priority.

Already tested, then the time is as before, tomorrow will check we return the exact same order of shapes descriptors.

Jetski5822 commented:
Well, I changed it so

if observer depends on subject, return 1.
if subject depends on observer, return -1.
if neither, return 0.

That file is not temp, its so I dont have a dependency, I might change it later, but for now its fine.

Nice idea on the shape table, if you want to fix and push, go for it... else I can try. let me know.

jtkech commented:
For compare i understand why my ordering tests was working, it's because when ordering we only use Compare() > 0`. Right?

Ok, finally i will do it now just after some little tests.

But anyway there is the stuff around views locations that i still need to do tomorrow.

jtkech commented:
Oh sorry for the error in `GetFeatureDependencies()` for testing because i've just done a copy paste on what was in `GetDependentFeatures()`. Need to sleep ;)

jtkech commented:
I've finished for the ordering improvement, will push soon.

jtkech commented:
@Jetski5822 

really need to leave, so last comment. About ordering using directives, sebastien said to me to always put System ones at the beginning.

jtkech commented:
@Jetski5822 

We can't use select but we can use just a regular OrderBy and it is much more faster.

It break one unit test because now we rely on the ordered list from the extension manager but in the unit tests we use a `TestExtensionManager` class, so maybe ... Will see tomorrow.

jtkech commented:
@Jetski5822 

Indeed, when testing shape table manager, because we now rely on the order of the list returned by `GetEnabledFeaturesAsync()`, the `TestExtensionManager` class need to return an ordered `_features` where the `DerivedTheme` is after the `BaseTheme`. I fixed it.

jtkech commented:
@Jetski5822 

So, i've pushed a 1st work to adapt the admin views expander. This to show the special case i want to reproduce from O1. And to be reviewed with @sebastienros to see if it is wrong.

I'll do more tests and give more details this night.

Here the admin expander has a lower priority than the theme expander. This because the theme expander naturally 1st implement the admin theme (and possible base themes). Then the admin expander, if it is the current theme, just add after all other enabled themes views locations.

Then, because of possible duplicates, i've used a `locations.Distinct()` in the composite expander.

Best.

jtkech commented:
@sebastienros @Jetski5822 

Just coming back to this after other works, and maybe i realized something. Maybe i was too concentrated to reproduce what is done in O1.

So i've reproduced what is done in O1 to allow a theme to override a controller view when in admin. For the admin theme name, the related code in O1 simply uses the string `"TheAdmin"`, but i assume that in O2 we can select another admin theme, so we needed to grab its name from the admin theme service.

But, precisely because now we can select a new admin theme (not in O1), why would we want to override admin mvc views in another way. For example, as i understand, to only override some views, now we can use a simple theme based on TheAdmin and then select it through the theme settings to be used in the admin.

**So, if i well understood, maybe we can just remove the new admin theme view expander**. If so you were right and now i know why.

jtkech commented:
Not so easy to return back to this after other works.

So, as said in my previous comment, because now we can create a new admin theme we can select, not sure it' worth to use an admin view expander to allow a non admin theme to override admin views.

That's said i just tried to override an admin controller view (could be a component view) without having to create a new admin theme (based or not on TheAdmin), just by using the following view in TheTheme.

    /Orchard2/src/Orchard.Cms.Web/Themes/TheTheme/Views/Orchard.Admin/Admin/Index.cshtml

Then, by using the admin view locations expander i got the following (see below). What's your feeling?

- **So, do you think it's worth to keep the admin view expander?**
- If so, i would like to know if we really need to use a "Shell" context value (not only the "Theme" one). We need it if enabled themes could be different per tenant.

I ask this because it seems that right now themes are installed / enabled through the setup recipe, then we can't enable / disable them, we can just select them. But maybe there is a missing ThemeService, or it is intended to use the ModuleService to enable / disable themes as modules, i don't know.

![admin expander](https://cloud.githubusercontent.com/assets/8586360/21206378/dd8b9d1a-c261-11e6-9900-359c3cf16865.png)

Best.

sebastienros commented:
Good that a base theme is enabled automatically because it becomes a dependency, I could like it.

Now, without handling the admin theme differently (just calling GetCurrentTheme)
- Create a new theme (TheAdmin2) based on TheAdmin and override a view there. Set it as the admin theme, ensure the view is used. Set TheAdmin as the admin theme, ensure the view is not used (without disabling TheAdmin2)
- Same thing for the front end theme.

Now bonus point: Create TheTheme2 which is not even based on another one, just a standard theme. Override a shape in this theme. Change the default theme to TheTheme2 and go on the front end to test the shape is used. Last time I checked there was an exception.

If it passes we are can merge and think about something else.

jtkech commented:
**About BaseThemes**

- I've checked where BaseTheme is specifically used in O1.

- In shape table manager but it is now ok with the updated code.

- In the ThemeService to enable / disable themes. But (see my previous comment) i could easily do it through the ModuleService as regular modules. The only specific thing in O1 is that when you disable a theme, the base theme is also disabled if not the current theme. But can be easily done if we need it.

- In WidgetService / PlacementService This is related to `Zones:` you can define in manifest files. But there is no more widgets zones in O2.

- Finally, here a base theme has the same impact on features ordering.

- So, **i didn't find any drawbacks to inject BaseTheme(s) in explicit dependencies**.

- Just one thing, in O1 it seems that we can have a theme without any assembly and which can still have a base theme. Maybe we will need to implement `Lightweight extensions` which are only defined through their manifest file but which can be enabled / disabled in a normal way. But can be done later.

**About multiple BaseThemes**

- Why not but when a derived theme doesn't override a view which is defined in 2 base themes we would need an order strategy. Maybe priorities but we would need to modify base themes manifest files. Or maybe just through the order they are declared in `BaseThemes` section. I let you decide on this part.

**About admin view expander**

- See my previous comment where i could easily override an admin controller view through the enabled theme TheTheme. And now i assume that the list of enabled themes depend on the tenant. So, the use of a Shell / Tenant contextual value (only if needed if we are in the admin).

Otherwise, on my side all in this PR is ok for me.

Best.

jtkech commented:
@sebastienros just seen your previous comment, not closed to my PC, will check all of you asked in 2 or 3 hours.

jtkech commented:
So, just began with the 1st test.

>Create a new theme (TheAdmin2) based on TheAdmin and override a view there. Set it as the admin theme, ensure the view is used. Set TheAdmin as the admin theme, ensure the view is not used (without disabling TheAdmin2)

- So, i only kept the `modules expander` and the `themes expander`, and inhibited the special `admin theme views locations expander`.

- Then, i've created `TheAdmin2` with only a `PlaceHolder.cs` and a `project.json`. Maybe good to be able to have an extension without any assembly.

- Then i've used `BaseThemes: TheAdmin` in the manifest, and added only the 2 following views.

      // a controller view
      /Orchard.Cms.Web/Themes/TheAdmin2/Views/Orchard.Admin/Admin/Index.cshtml
      // a shape
      /Orchard.Cms.Web/Themes/TheAdmin2/Views/Navigation-admin.cshtml

- Then i've enabled `TheAdmin2` through the `ModuleService` where i've removed the themes filtering.

- Then when you change the admin theme selection, the controller view changes immediately, but for the shape (here `Navigation-admin`) you need to wait about 1 minute without refreshing to get the new view rendering. Maybe some shape caching somewhere, otherwise it works.

![theadmin2](https://cloud.githubusercontent.com/assets/8586360/21243752/7f2cc88e-c31a-11e6-8e7c-6776c7eed3d9.png)

So, i think you will say that we can get rid of the `admin theme view locations expander` and just use the regular `theme expander` which only use the current theme. This because if we can create a simple admin theme base on TheAdmin, we don't need to override an admin view from another theme as in O1. Right?

Will continue with the 2 other tests.

Best.

jtkech commented:
2nd test.

>Same thing for the front end theme.

It works with the same remark as in the 1st test. Here i needed to wait a little more time (maybe 5 minutes) without refreshing the page, this to get the right shape rendering. For a controller view (here i override the Orchard.Demo home index view) we don't need to wait. Otherwise it works as expected.

Todo: the following last test.

>Now bonus point: Create TheTheme2 which is not even based on another one ...

jtkech commented:
Last test.

>Now bonus point: Create TheTheme2 which is not even based on another one, just a standard theme. Override a shape in this theme. Change the default theme to TheTheme2 and go on the front end to test the shape is used. Last time I checked there was an exception.

- It works. Note: here, after switching the theme, i don't need to wait to get the right shape rendering.

![thetheme2](https://cloud.githubusercontent.com/assets/8586360/21246515/970a3f68-c329-11e6-9bb6-3f47cfbe468b.png)

Jetski5822 commented:
So all scenarios work?

Sent from my iPhone

> On 16 Dec 2016, at 07:56, jtkech <notifications@github.com> wrote:
> 
> Last test.
> 
> Now bonus point: Create TheTheme2 which is not even based on another one, just a standard theme. Override a shape in this theme. Change the default theme to TheTheme2 and go on the front end to test the shape is used. Last time I checked there was an exception.
> 
> It works. Note: here, after switching the theme, i don't need to wait to get the right shape rendering.
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jtkech commented:
@sebastienros @Jetski5822 

- Yes, the above tests passed. So, i think you want that i remove `AdminThemeAwareViewLocationExpanderProvider`. **Will do it in a few minutes**.

- About having multiples base themes (here not through the hierarchy), maybe we don't need to worry about ordering. If 2 themes define many of the same views, it's not good to use both as base themes.

- But, if possible, the other useful usage i see is to use more specialized themes. E.g Theme1 which defines some kind of views, Theme2 other kind of views (not the same views). Then a global theme which has Theme1 and Theme2 as base themes.

Jetski5822 commented:
If you could remove that would be awesome :)

Sent from my iPhone

> On 16 Dec 2016, at 08:35, jtkech <notifications@github.com> wrote:
> 
> @sebastienros @Jetski5822
> 
> Yes, the above tests passed. So, i think you want that i remove AdminThemeAwareViewLocationExpanderProvider. Will do it in a few minutes.
> 
> About having multiples base themes (here not through the hierarchy), maybe we don't need to worry about ordering. If 2 themes define many of the same views, it's not good to use both as base themes.
> 
> But, if possible, the other useful usage i see is to use more specialized themes. E.g Theme1 which defines some kind of views, Theme2 other kind of views (not the same views). Then a global theme which has Theme1 and Theme2 as base themes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sebastienros commented:
I would like the multiple base themes to go away. I can't even explain how it would work ;) Please @Jetski5822

jtkech commented:
@Jetski5822 yes i'll remove it.

But please wait before doing anything else because just seen another use case. During my tests i was using TheTheme2, then i've deleted it from the disk but still selected in the database, then when i re-start the app it can't find TheTheme2 and fails. So maybe here another useful null check to do!

Jetski5822 commented:
Let's have a chat about the multiple base themes.

Sent from my iPhone

> On 16 Dec 2016, at 08:51, jtkech <notifications@github.com> wrote:
> 
> @Jetski5822 yes i'll remove it.
> 
> But please wait before doing anything else because just seen another use case. During my tests i was using TheTheme2, then i've deleted it from the disk but still selected in the database, then when i re-start the app it can't find TheTheme2 and fails. So maybe here another useful null check to do!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sebastienros commented:
@Jetski5822 convince me ;)

jtkech commented:
@sebastienros 

Please wait i will remove admin view expander

sebastienros commented:
He can still try to convince me while you remove stuff.

jtkech commented:
So, after removing TheTheme2 from the disk which was still selected in the database. I could restart the app by adding some null checks in ThemeManager, ExtensionManager and in the Theme view expander.

Because quickly done, if you can review it. Will push it soon.

Otherwise, just checked that all unit tests pass. Just want to re-run a new setup.

jtkech commented:
Is there anything else i could do here?

@sebastienros when you said

>@Jetski5822 convince me ;)

I understood that he had convinced you.

@Jetski5822 when you said

>Let's have a chat about the multiple base themes.

I understood with Sebastien. If with me, feel free to ping me.

----
As a side note i've tried to use a theme extension which is not a project.

- I've copied `TheTheme` in `TheTheme2` but without any `PlaceHolder.cs` and no project json files, so it's not a project and without any assembly. So, we try to load it as a precompiled extension ...

- Then, by checking that there is no binary folder, i could prevent an exception to be thrown.

- Then it works, i think because of things we can inject through razor files by using `@using`, `@inherits` ...

Maybe worth to implement extensions which are not projects (i mainly think about themes).

Jetski5822 commented on src/Orchard.DisplayManagement/Theming/ThemeManager.cs:
```diff
@@ -47,7 +47,7 @@ public class ThemeManager : IThemeManager
                 {
                     var t = _extensionManager.GetExtension(theme.ThemeName);
-                    if (t.Exists)
+                    if (t?.Exists ?? false)
```
@jtkech Can you revert this, Get Extension should return a NotFoundExtension class, with Exists = False. Get Extension should not return null.

jtkech replied:
Ok. So here it was to prevent from an exception when the current selected theme was no more present on the disk. So, ok if `GetExtension()` returns a non null object. Will revert it this night asap.

Jetski5822 replied:
I got time, I will take care of it now

jtkech replied:
Perfect, thanks

Jetski5822 commented on src/Orchard.DisplayManagement/LocationExpander/ThemeAwareViewLocationExpanderProvider.cs:
```diff
@@ -23,7 +23,12 @@ public void PopulateValues(ViewLocationExpanderContext context)
-                context.Values["Theme"] = themeManager.GetThemeAsync().GetAwaiter().GetResult().Id;
+                var currentTheme = themeManager.GetThemeAsync().GetAwaiter().GetResult();
+
+                if (currentTheme != null)
```
@jtkech maybe introduce currentTheme.Exists ?

Jetski5822 commented on src/Orchard.Environment.Extensions/ExtensionManager.cs:
```diff
@@ -82,7 +82,12 @@ public IExtensionInfo GetExtension(string extensionId)
+                return _extensionsById[extensionId];
+            }
+
+            return null;
```
NotFoundExtensionInfo ?

jtkech replied:
Ok. Here also, when the current selected theme is no more present on the disk, it was throwing an exception because the dictionary key doesn't exist.

Ok but didn't find a `NotFoundExtensionInfo` class already defined. So, do you mean i have to create such a class (e.g in `Extensions.Abstractions`)? Which embed e.g a `NotFoundFileInfo` and a `NotFoundManifestInfo`. Right? If so, will do it this night asap.

Jetski5822 commented:
Okay, can some one approve this now? @sebastienros @agriffard  ??

sebastienros commented:
No, still waiting for the base themes explanation or removal.

On Dec 17, 2016 7:47 AM, "Nicholas Mayne" <notifications@github.com> wrote:

> Okay, can some one approve this now? @sebastienros
> <https://github.com/sebastienros> @agriffard
> <https://github.com/agriffard> ??
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrchardCMS/Orchard2/pull/402#issuecomment-267769661>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABHJ7TWWediXzQCeKWKut30RHdxCNmzLks5rJAQqgaJpZM4LJsl0>
> .
>

Jetski5822 commented:
Ah yeah forgot about that...

jtkech commented:
@Jetski5822 

About `NotFoundExtension` i was beginning to do quite the same ;)

I learned a lot working with you. No luck here, each time i was doing something, i was breaking another thing. But i was more concentrated that all things are taken into account (e.g implicit dependencies of themes on modules), even if it was failing at the beginning.

So, i hope i didn't disturb you too much.

That's said, if `HasDependency()` is only used by the `OrderByDependenciesAndPriorities()` utility, as far as i can test, with the following code (as in O1) we get the same (because the utility do the job) and it is faster. But i agree with you if you think it can be used in another context (and because now it is cached). But here i would like to know your real reason.


        private bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
        {
            //return _extensionOrderingStrategy.Compare(f1, f2) > 0 ? true : GetFeatureDependencies(f1.Id).Contains(f2);
            return _extensionOrderingStrategy.Compare(f1, f2) > 0 ? true : f1.Dependencies.Contains(f2.Id);
        }

Need to leave.

Best.

Jetski5822 commented:
GetFeatureDependencies is the true list of dependencies for a given
feature, the f1.Dependencies only lists what is in the Manifest, and
therefore is just a list of immediate dependencies, and not the true
dependency list.

Let me see if I can write some crazy tests tomorrow to break it and if it
doesnt break we will use the Property rather than the method.

N

On Sun, Dec 18, 2016 at 12:47 AM, jtkech <notifications@github.com> wrote:

> @Jetski5822 <https://github.com/Jetski5822>
>
> About NotFoundExtension i was beginning to do quite the same ;)
>
> I learned a lot working with you. No luck here, each time i was doing
> something, i was breaking another thing. But i was more concentrated that
> all things are taken into account (e.g implicit dependencies of themes on
> modules), even if it was failing at the beginning.
>
> So, i hope i didn't disturb you too much.
>
> That's said, if HasDependency() is only used by the
> OrderByDependenciesAndPriorities() utility, as far as i can test, with
> the following code (as in O1) we get the same (because the utility do the
> job) and it is faster. But i agree with you if you think it can be used in
> another context (and because now it is cached). But here i would like to
> know your real reason.
>
>     private bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
>     {
>         //return _extensionOrderingStrategy.Compare(f1, f2) > 0 ? true : GetFeatureDependencies(f1.Id).Contains(f2);
>         return _extensionOrderingStrategy.Compare(f1, f2) > 0 ? true : f1.Dependencies.Contains(f2.Id);
>     }
>
> Need to leave.
>
> Best.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrchardCMS/Orchard2/pull/402#issuecomment-267772749>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAl0ChAJSAvI5MyrXp9agebEJqREB5dcks5rJBILgaJpZM4LJsl0>
> .
>

jtkech commented:
Ok, thanks for your patience.

And in your tests, i assume you use it to order a collection of features which also contains all their hierarchical dependencies. I mean if f1 depends on f2 and f2 depends on f3, i assume that f1, f2 and f3 are all in the collection we are ordering.

>GetFeatureDependencies is the true list of dependencies

You're right but it seems that the ordering utility also go through the graph in some way. For a given item the utility retrieve its dependencies (with HasDependency) in the population (to add them before), but before adding the item it also recursively do the same for each of these dependencies, and so on.

So, it seems that we get the same result if, as said above, all hierarchical dependencies of each feature also are in the collection we are ordering.

Best.

jtkech commented:
@Jetski5822 just for infos, after a little reading on this, i found that our ordering utility uses a kind of the `Depth-first search` algorithm used for topological sorting. Normally, for each node, it only needs to know direct dependencies (tree edges, not forward edges) which are parsed recursively through the graph.

Best.

jtkech commented:
>Let me see if I can write some crazy tests tomorrow to break it and if it
doesnt break we will use the Property rather than the method.

Did you check it? Note: i always knew that the property is only immediate dependencies. But this is what needs `OrderByDependenciesAndPriorities()`. See in the code how `Add()` uses `hasDependency` and is recursively called to add dependencies of dependencies of dependencies ... This before the item itself.

Maybe good to rename `HasDependency` to e.g `HasDirectDependency` as it is used by the ordering utility.

Need more time but i also want to open an "issue" to discuss on little points (not only this one). I think all is ok, but to be sure we don't miss anything (e.g related to the special "core feature" which provides shapes).

Best.
